src/libraries/System.Private.CoreLib/src/System/MemoryExtensions.cs

```diff
@@ -2009,6 +2009,102 @@ public static void Sort<T>(this Span<T> span, Comparison<T> comparison)
+        /// <typeparam name="T">The type of the elements in the spans.</typeparam>
+        /// <param name="span">The first sequence to compare.</param>
+        /// <param name="other">The second sequence to compare.</param>
+        /// <returns>The length of the common prefix shared by the two spans.  If there's no shared prefix, 0 is returned.</returns>
+        public static int CommonPrefixLength<T>(this Span<T> span, ReadOnlySpan<T> other) =>
+            CommonPrefixLength((ReadOnlySpan<T>)span, other);
+
+        /// <summary>Finds the length of any common prefix shared between <paramref name="span"/> and <paramref name="other"/>.</summary>
+        /// <typeparam name="T">The type of the elements in the spans.</typeparam>
+        /// <param name="span">The first sequence to compare.</param>
+        /// <param name="other">The second sequence to compare.</param>
+        /// <param name="comparer">The <see cref="IEqualityComparer{T}"/> implementation to use when comparing elements, or null to use the default <see cref="IEqualityComparer{T}"/> for the type of an element.</param>
+        /// <returns>The length of the common prefix shared by the two spans.  If there's no shared prefix, 0 is returned.</returns>
+        public static int CommonPrefixLength<T>(this Span<T> span, ReadOnlySpan<T> other, IEqualityComparer<T>? comparer) =>
+            CommonPrefixLength((ReadOnlySpan<T>)span, other, comparer);
+
+        /// <summary>Finds the length of any common prefix shared between <paramref name="span"/> and <paramref name="other"/>.</summary>
+        /// <typeparam name="T">The type of the elements in the spans.</typeparam>
+        /// <param name="span">The first sequence to compare.</param>
+        /// <param name="other">The second sequence to compare.</param>
+        /// <returns>The length of the common prefix shared by the two spans.  If there's no shared prefix, 0 is returned.</returns>
+        public static int CommonPrefixLength<T>(this ReadOnlySpan<T> span, ReadOnlySpan<T> other)
+        {
+            // Shrink one of the spans if necessary to ensure they're both the same length. We can then iterate until
+            // the Length of one of them and at least have bounds checks removed from that one.
+            if (other.Length > span.Length)
+            {
+                other = other.Slice(0, span.Length);
+            }
+            else if (span.Length > other.Length)
+            {
+                span = span.Slice(0, other.Length);
+            }
+            Debug.Assert(span.Length == other.Length);
+
+            // Find the first element pairwise that is not equal, and return its index as the length
+            // of the sequence before it that matches.
+            for (int i = 0; i < span.Length; i++)
+            {
+                if (!EqualityComparer<T>.Default.Equals(span[i], other[i]))
+                {
+                    return i;
+                }
+            }
```

Comment:
This could be vectorized for bit-wise comparable types (as well below when the default comparer is used).
Is this something to look in (in a different PR, to have this change here concise)?

Reply:
We can certainly consider it.  My intuition (which could be wrong, of course) is prefixes will be relatively short, in which case I'd want to favor minimal overhead over maximal throughput.  If there's a reasonably small implementation that's maintainable, minimal, and little overhead in the expected short cases while still providing benefit, I'd be game.

(If you're interested in vectorizing something, I'd personally be much more interested in seeing investment in https://github.com/dotnet/runtime/issues/67942.)

Reply:
Once the PRs (this, and https://github.com/dotnet/runtime/issues/67942) are merged, I'll have a look, considering https://github.com/dotnet/runtime/issues/67942 first.

> common prefixes will be relatively short

I don't know 😉 But with `Vector128<ushort>` (for char) it's 8 chars and vectorization can be done. So my feeling* is that this threshould is hit quite often. Of course I could be wrong here too.
I'll try it and see if the additional complexity is worth it.

\* based on a use case for an (internal) parser I have in mind

Reply:
Regardless, I'm glad there's interest in consuming the API in general :)